warning.rb

```diff
@@ -44,10 +44,16 @@ module Kernel
   # following categories:
   #
   # :deprecated :: Used for warning for deprecated functionality that may
-  #                be removed in the future.
+  #                  be removed in the future.
   # :experimental :: Used for experimental features that may change in
   #                  future releases.
+  # :performance  :: Used for warning about APIs or pattern that have
+  #                  negative performance impact
   def warn(*msgs, uplevel: nil, category: nil)
-    Primitive.rb_warn_m(msgs, uplevel, category)
+    if category.nil?
```

Comment:
This works unexpectedly if `category.nil?` is overridden.

```suggestion
    if Primitive.cexpr!("NIL_P(category)")
```